x-pack/plugins/aiops/server/routes/queries/get_simple_hierarchical_tree.ts

```diff
@@ -0,0 +1,267 @@
+
+  return keys.sort((a, b) => {
+    return valueCounts[b] - valueCounts[a];
+  });
+}
+
+interface NewNode {
+  name: string;
+  set: FieldValuePair[];
+  docCount: number;
+  children: NewNode[];
+  icon: string;
+  iconStyle: string;
+  addNode: (node: NewNode) => void;
+}
+
+function NewNodeFactory(name: string): NewNode {
+  const children: NewNode[] = [];
+
+  const addNode = (node: NewNode) => {
+    children.push(node);
+  };
+
+  return {
+    name,
+    set: [],
+    docCount: 0,
+    children,
+    icon: 'default',
+    iconStyle: 'default',
+    addNode,
+  };
+}
+
+/**
+ * Simple (poorly implemented) function that constructs a tree from an itemset DataFrame sorted by support (count)
+ * The resulting tree components are non-overlapping subsets of the data.
+ * In summary, we start with the most inclusive itemset (highest count), and perform a depth first search in field order.
+ *
+ * TODO - the code style here is hacky and should be re-written
+ *
+ * @param displayParent
+ * @param parentDocCount
+ * @param parentLabel
+ * @param field
+ * @param value
+ * @param iss
+ * @param collapseRedundant
+ * @param displayOther
+ * @returns
+ */
+function dfDepthFirstSearch(
+  fields: string[],
+  displayParent: NewNode,
+  parentDocCount: number,
+  parentLabel: string,
+  field: string,
+  value: string,
+  iss: ItemsetResult[],
+  collapseRedundant: boolean,
+  displayOther: boolean
+) {
+  const filteredItemSets = iss.filter((is) => {
+    for (const [key, values] of Object.entries(is.set)) {
+      if (key === field && values.includes(value)) {
+        return true;
+      }
+    }
+    return false;
+  });
+
+  if (filteredItemSets.length === 0) {
+    return 0;
+  }
+
+  const docCount = max(filteredItemSets.map((fis) => fis.doc_count)) ?? 0;
+  const totalDocCount = max(filteredItemSets.map((fis) => fis.total_doc_count)) ?? 0;
+
+  let label = `${parentLabel} ${value}`;
+
+  let displayNode: NewNode;
+  if (parentDocCount === docCount && collapseRedundant) {
+    // collapse identical paths
+    displayParent.name += ` ${value}`;
+    displayParent.set.push({ fieldName: field, fieldValue: value });
+    displayParent.docCount = docCount;
+    displayNode = displayParent;
+  } else {
+    displayNode = NewNodeFactory(`${docCount}/${totalDocCount}${label}`);
+    displayNode.iconStyle = 'warning';
+    displayNode.set = [...displayParent.set];
+    displayNode.set.push({ fieldName: field, fieldValue: value });
+    displayNode.docCount = docCount;
+    displayParent.addNode(displayNode);
+  }
+
+  let nextField: string;
+  while (true) {
+    const nextFieldIndex = fields.indexOf(field) + 1;
+    if (nextFieldIndex >= fields.length) {
+      displayNode.icon = 'file';
+      displayNode.iconStyle = 'info';
+      return docCount;
+    }
+    nextField = fields[nextFieldIndex];
+
+    // TODO - add handling of creating * as next level of tree
+
+    if (Object.keys(getValueCounts(filteredItemSets, nextField)).length > 0) {
```

Comment:
Any chance the break condition is never hit and we stay in an infinite while loop?